Issue #3650: Expand missing testing for PropertyCacheFile tests

MEZk opened the pull request:
#3650

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:
```diff
@@ -418,7 +420,7 @@ public void testDestroyCacheWithWrongFileNameLength() throws Exception {
 
     @Test
-    public void testCacheFile() throws Exception {
+    public void testCacheFileWithoutExternalResources() throws Exception {
```
I am not sure what `WithoutExternalResources` means in terms of the expected code or goal of the test. Can you explain?

MEZk replied:
The name of the test means that we test caching functionality for the check which does not hold external resources.

rnveach replied:
> we test caching functionality for the check which does not hold external resources.

So this is the same as `testFilterWhichDoesNotImplementExternalResourceHolderInterface` but that is testing it for a filter and this one is testing it with a check?

MEZk replied:
Yes, I think that it might be called as testCheckWhichDoesNotImplementExternalResourceHolderInterface. Are you Ok with the name? Methods can be grouped closer.

rnveach replied:
Thanks, that clears things up.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:
```diff
@@ -455,36 +466,44 @@ public void testCacheFileChangeInConfig() throws Exception {
+        assertNotEquals(
+            cacheAfterFirstRun.getProperty("configuration*?"),
+            cacheAfterSecondRun.getProperty("configuration*?")
+        );
```
Can we add an assertion that we have the same number of properties in the cache?

MEZk replied:
Done.

MEZk commented on src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:
```diff
@@ -526,37 +543,37 @@ public void testClearExistingCache() throws Exception {
-        checkerConfig.addAttribute("charset", "UTF-8");
-        checkerConfig.addChild(treeWalkerConfig);
-
+        // The idea of the test is to check that clearing of the non-existent cache
```
Renamed to ```testClearCacheWhenCacheFileIsNotSet``` as ```testClearCacheWithNoCache``` is a bit confusing due to ```Cache...NoCache```.

MEZk commented on src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:
```diff
@@ -589,42 +606,76 @@ public void testCatchErrorInProcessFilesMethod() throws Exception {
         // We invoke 'verify' twice to invalidate cache
         // and have two identical exceptions which happen on the same line between runs
         final int numberOfRuns = 2;
+        final String[] md5AfterFirstAndSecondRuns = new String[numberOfRuns];
```
Done.

MEZk commented on src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:
```diff
@@ -589,42 +606,76 @@ public void testCatchErrorInProcessFilesMethod() throws Exception {
+        final String[] md5AfterFirstAndSecondRuns = new String[numberOfRuns];
+        final Properties cache = new Properties();
         for (int i = 0; i < numberOfRuns; i++) {
             verify(mockChecker, pathToEmptyFile, expected);
```
I cannot figure out how to intersect the exception. I could generate md5 of expected exception manually and assert that actual md5 after first and second run is equal to expected, however, if we change the line of ```verify``` method invocation the md5 will be changed.

rnveach replied:
> if we change the line of verify 

Not just `verify` but anything that is in the stack trace too. So that won't work.

> I cannot figure out how to intersect the exception

I can't see anything either unless we use powermock to force a specific exception that we control.

> Do you mean that we should make sure that getting non_existing_external_resource.xml before the invocation of verify method in the loop throws an exception

That sounds like what we will have to do.

MEZk replied:
Done. The test was moved to PropertyCacheFile. See ```testPutNonExsistingExternalResourceDifferentExceptionsBetweenRuns```.

MEZk commented on src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:
```diff
@@ -589,42 +606,76 @@ public void testCatchErrorInProcessFilesMethod() throws Exception {
         // Once again to invalidate cache because in the second run exception will happen
-        // on different line
+        // on the different line
         verify(mockChecker, pathToEmptyFile, expected);
```
answered in the previous comment.

MEZk commented on src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:
```diff
@@ -657,9 +731,24 @@ public void testFilterWhichDoesNotImplementExternalResourceHolderInterface() thr
+        assertEquals(
+            cacheAfterFirstRun.getProperty("configuration*?"),
+            cacheAfterSecondRun.getProperty("configuration*?")
+        );
```
Done.

MEZk commented on src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:
```diff
@@ -589,42 +606,76 @@ public void testCatchErrorInProcessFilesMethod() throws Exception {
 
     @Test
-    public void testMultipleConfigs() throws Exception {
+    public void testMultipleCheckConfigs() throws Exception {
```
Removed.

MEZk commented:
@rnveach 
Thank you for the review. Could you review the changes once again since I removed one test and moved 2 tests from CheckerTest to PropertyCacheFile?

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:
```diff
@@ -418,7 +419,8 @@ public void testDestroyCacheWithWrongFileNameLength() throws Exception {
 
     @Test
-    public void testCacheFile() throws Exception {
+    public void testCheckWhichDoesNotImplementExternalResourceHolderInterface() throws Exception {
```
put `cache` in here somewhere since main point is checking cache.

MEZk replied:
Done

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:
```diff
@@ -494,12 +463,21 @@ public void testWithCacheWithNoViolation() throws Exception {
+        cache.load(Files.newBufferedReader(cacheFile.toPath()));
+
+        // There should be no objects in cache since there are no violations and config is not set.
+        assertTrue(cache.isEmpty());
```
I'm not sure if I didn't notice this before, but this test is a lie.
`PropertyCacheFile` can't be instantiated without a configuration, it will throw an exception.
The whole reason the cache is empty is because `destroy` is never called to commit the changes because we play with `Checker` directly instead of going through a 'normal' lifecycle.

We specifically set the configuration with the line `checker.configure(createCheckConfig(TranslationCheck.class));`. So it is a lie there is no configuration.
Also if there are no violations, cache is still filled with all files checked.

Is this test relying on some code coverage? If so, let's rework this test around that.

MEZk replied:
Added 
```java
// invoke destroy to persist cache
checker.destroy();
```

to persist cache. Changed asserts.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:
```diff
@@ -526,37 +502,37 @@ public void testClearExistingCache() throws Exception {
 
         verify(checker, pathToEmptyFile, pathToEmptyFile, expected);
+
         checker.clearCache();
```
why call `verify` if we are going to clear cache right after before checking anything?

MEZk replied:
Removed ```verify```.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:
```diff
@@ -588,67 +564,14 @@ public void testCatchErrorInProcessFilesMethod() throws Exception {
-    }
-
-    @Test
     public void testFilterWhichDoesNotImplementExternalResourceHolderInterface() throws Exception {
```
Same comment as before, add `cache` somewhere.

MEZk replied:
Done

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/CheckerTest.java:
```diff
@@ -672,11 +613,13 @@ public void testCheckAddsNewResourceLocationButKeepsSameCheckerInstance() throws
         // Cache should not be reused.
```
Can't comment on method name, but same comment as others, add cache somewhere to method name.

MEZk replied:
Done.

rnveach commented:
@MEZk ping

MEZk commented:
@rnveach 
Working on it.

MEZk commented:
@rnveach 
> In CheckerTest, the following are not used anymore. Either find them a use or remove them.
createMockCheckerWithCacheForModule, DummyFileSetCheck, MockMissingExternalResourcesFileSetCheck

Removed. Please, review.

romani commented:
@MEZk , TC issues could be resolved by suppressions in javadoc. I am ok with long Test methods.

Travis is restarted, if it fail again - it should be fixed.

MEZk commented:
@romani 
>  TC issues could be resolved by suppressions in javadoc. I am ok with long Test methods.

Fixed.

romani commented:
@MEZk , thanks a lot , please do not spend time on such issues any more, it will be addressed during pitest coverage.
